lib/ed25519.js

```diff
@@ -0,0 +1,148 @@
+ *
+ * @param {Buffer} buffer - A DER encoded key buffer.
+ *
+ * @throws {Error} If the buffer does not contain a valid DER Prefix.
+ *
+ * @returns {Buffer} The key material part of the Buffer.
+*/
+function getKeyMaterial(buffer) {
+  if(buffer.indexOf(DER_PUBLIC_KEY_PREFIX) === 0) {
+    return buffer.slice(DER_PUBLIC_KEY_PREFIX.length, buffer.length);
+  }
+  if(buffer.indexOf(DER_PRIVATE_KEY_PREFIX) === 0) {
+    return buffer.slice(DER_PRIVATE_KEY_PREFIX.length, buffer.length);
+  }
+  throw new Error('Expected Buffer to match Ed25519 Public or Private Prefix');
+}
+
+/**
+ * Takes a Buffer or Uint8Array with the raw private key and encodes it
+ * in DER-encoded PKCS#8 format.
+ * Allows Uint8Arrays to be interoperable with node's crypto functions.
+ *
+ * @param {object} options - Options to use.
+ * @param {Buffer} [options.privateKeyBytes] - Required if no seedBytes.
+ * @param {Buffer} [options.seedBytes] - Required if no privateKeyBytes.
+ *
+ * @throws {TypeError} Throws if the supplied buffer is not of the right size
+ *  or not a Uint8Array or Buffer.
+ *
+ * @returns {Buffer} DER private key prefix + key bytes.
+*/
+function privateKeyDerEncode({privateKeyBytes, seedBytes}) {
+  if(!(privateKeyBytes || seedBytes)) {
+    throw new TypeError('`privateKeyBytes` or `seedBytes` is required.');
+  }
+  if(!privateKeyBytes && !(seedBytes instanceof Uint8Array &&
+    seedBytes.length === 32)) {
+    throw new TypeError('`seedBytes` must be a 32 byte Buffer.');
+  }
+  if(!seedBytes && !(privateKeyBytes instanceof Uint8Array &&
+    privateKeyBytes.length === 64)) {
+    throw new TypeError('`privateKeyBytes` must be a 64 byte Buffer.');
+  }
+  let p;
+  if(seedBytes) {
+    p = seedBytes;
+  } else {
+    // extract the first 32 bytes of the 64 byte private key representation
+    p = privateKeyBytes.slice(0, 32);
+  }
+  return Buffer.concat([DER_PRIVATE_KEY_PREFIX, p]);
+}
+
+/**
+ * Takes a Buffer of public key bytes and encodes it in DER-encoded
+ * SubjectPublicKeyInfo (SPKI) format.
+ * Allows Uint8Arrays to be interoperable with node's crypto functions.
+ *
+ * @param {object} options - Options to use.
+ * @param {Uint8Array} options.publicKeyBytes - The keyBytes.
+ *
+ * @throws {TypeError} Throws if the bytes are not Uint8Array or of length 32.
+ *
+ * @returns {Buffer} DER Public key Prefix + key bytes.
+*/
+function publicKeyDerEncode({publicKeyBytes}) {
+  if(!(publicKeyBytes instanceof Uint8Array && publicKeyBytes.length === 32)) {
+    throw new TypeError('`publicKeyBytes` must be a 32 byte Buffer.');
+  }
+  return Buffer.concat([DER_PUBLIC_KEY_PREFIX, publicKeyBytes]);
+}
+
+// seedBytes can be a TypedArray or a function that returns a TypedArray
+const getSeedBytes = async seed => {
+  if(typeof seed === 'function') {
+    return seed();
+  }
+  return seed;
+};
+
+const api = {
+  async generateKeyPairFromSeed(seed) {
+    const seedBytes = await getSeedBytes(seed);
```

Comment:
We shouldn't call `getSeedBytes` but instead validate that `seed` is a 32-byte Uint8Array.

Reply:
in order to be backwards compatible with `stablelib/ed25519` we do. It might be better to accept that seed is a function if we want to simplify.

Reply:
We're miscommunicating. First, we don't need to be "backwards compatible" with `stablelib/ed25519` because it is just an implementation detail. Second, we should change the internal module interface that is split between the browser/node to do what we need; we don't have to code directly to `stablelib/ed25519`'s API if that causes some trouble, which it seems to be doing here -- it's spilling details that should be hidden in the browser version into the code that is used in both the browser and node.

Reply:
I believe this address this issue: https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/11d6eb9cf0ce47777e6fe4ec4e6386a922b94ca5